lib/kino/js/live.ex

```diff
@@ -0,0 +1,369 @@
+  ## Event handlers
+
+  You must eventually register JavaScript handlers for all events
+  that the client may receive. However, the registration can be
+  deferred, if the initialization is asynchronous. For example,
+  the following is perfectly fine:
+
+  ```js
+  export function init(ctx, data) {
+    fetch(data.someUrl).then((resp) => {
+      ctx.handleEvent("update", (payload) => {
+        // ...
+      });
+    });
+  }
+  ```
+
+  Or alternatively:
+
+  ```js
+  export async function init(ctx, data) {
+    const response = await fetch(data.someUrl);
+
+    ctx.handleEvent("update", (payload) => {
+      // ...
+    });
+  }
+  ```
+
+  In such case all incoming events are buffered and dispatched once
+  the handler is registered.
+  '''
+
+  defstruct [:module, :pid]
+
+  @doc false
+  use GenServer
+
+  require Logger
+
+  alias Kino.JS.Live.Context
+
+  @type t :: %__MODULE__{module: module(), pid: pid()}
+
+  @doc """
+  Invoked when the widget server started.
+
+  See `c:GenServer.init/1` for more details.
+  """
+  @callback init(arg :: term(), ctx :: Context.t()) :: {:ok, ctx :: Context.t()}
+
+  @doc """
+  Invoked whenever a new client connects to the server.
+
+  The returned data is passed to the JavaScript `init` function
+  of the connecting client.
+  """
+  @callback handle_connect(ctx :: Context.t()) :: {:ok, data :: term(), ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle client events.
+  """
+  @callback handle_event(event :: String.t(), payload :: term(), ctx :: Context.t()) ::
+              {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle asynchronous `cast/2` messages.
+
+  See `c:GenServer.handle_cast/2` for more details.
+  """
+  @callback handle_cast(msg :: term(), ctx :: Context.t()) :: {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle synchronous `call/3` messages.
+
+  See `c:GenServer.handle_call/3` for more details.
+  """
+  @callback handle_call(msg :: term(), {pid(), reference()}, ctx :: Context.t()) ::
+              {:noreply, ctx :: Context.t()} | {:reply, term(), ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle all other messages.
+
+  See `c:GenServer.handle_info/2` for more details.
+  """
+  @callback handle_info(msg :: term(), ctx :: Context.t()) :: {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked when the server is about to exit.
+
+  See `c:GenServer.terminate/2` for more details.
+  """
+  @callback terminate(reason, ctx :: Context.t()) :: term()
+            when reason: :normal | :shutdown | {:shutdown, term()}
+
+  @optional_callbacks init: 2,
+                      handle_event: 3,
+                      handle_call: 3,
+                      handle_cast: 2,
+                      handle_info: 2,
+                      terminate: 2
+
+  defmacro __using__(opts) do
+    quote location: :keep, bind_quoted: [opts: opts] do
+      @behaviour Kino.JS.Live
+
+      import Kino.JS.Live.Context, only: [assign: 2, update: 3, broadcast_event: 3]
```

Comment:
I would rather flip things around. Keep `assign/2`, `update/3` and user facing API in `Kino.JS.Live` and have the server in `Kino.JS.LiveServer`. This way we reduce the surface API for users.

Reply:
Sounds good, let me know if I interpreted it right c8345e2e8420a9e81ab698f0249e63569e12e706.

Reply:
Yes, exactly. Next I would remove `use Live` altogether.

Reply:
The `@behaviour` + `import` combination is gonna be pretty much the default, shouldn't we make this scenario simpler?

Reply:
Ok then. Should it use Kino.JS automatically as well?

Reply:
That's what I thought initially, but `Kino.JS` accepts options (`:asset_path`) and option forwarding feels too indirect.